hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java

```diff
@@ -64,29 +66,33 @@ public interface TokenBalanceRepository extends CrudRepository<TokenBalance, Tok
     @Query(
             value =
                     """
-                    with balance_snapshot as (
-                      select consensus_timestamp
-                      from account_balance
-                      where account_id = 2 and consensus_timestamp <= ?3
-                      order by consensus_timestamp desc
-                      limit 1
-                    ), base as (
-                        select balance
-                        from token_balance as tb, balance_snapshot as s
-                        where
-                          tb.consensus_timestamp = s.consensus_timestamp and
-                          token_id = ?1 and
-                          account_id = ?2
-                    ), change as (
-                        select sum(amount) as amount
-                        from token_transfer as tt, balance_snapshot as s
-                        where
-                          token_id = ?1 and
-                          account_id = ?2 and
-                          tt.consensus_timestamp > s.consensus_timestamp and
-                          tt.consensus_timestamp <= ?3
-                    )
-                    select coalesce((select balance from base), 0) + coalesce((select amount from change), 0)
+                        with balance_timestamp as (
+                        select consensus_timestamp
+                        from account_balance
+                        where account_id = 2 and
+                            consensus_timestamp > ?3 - 2678400000000000 and consensus_timestamp <= ?3
+                        order by consensus_timestamp desc
+                        limit 1
+                        ), base as (
+                            select tb.balance, tb.consensus_timestamp
```

Comment:
```suggestion
                            select tb.balance
```
Using the token's balance timestamp when token balance info is deduped can increase the timestamp range when querying token_transfer table and it's completely unnecessary.

For example, based on account 2, the balance timestamp is 5000. In the `base` CTE, the latest deduped token balance info at or before timestamp 5000 has `balance = 20, consensus_timestamp = 1000`.

This indicates for all the balance snapshot timestamps in the range `[1000,  5000]`, the token's. balance is always 20. To reconcile and find the token balance change, we should query token_balance table in the timestamp range `(5000,  blockTimestamp]`,  instead of `(1000, blockTimestamp]`, because the deduped balance info implies in the timestamp range `(1000, 5000]`, there is no token transfer for the account and token at all.

Reply:
Fixed